File: robocode.samples/src/main/java/sample/Walls.java
/*
 * Copyright (c) 2001-2025 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
package sample;


import robocode.HitRobotEvent;
import robocode.Robot;
import robocode.ScannedRobotEvent;

import java.awt.*;


/**
 * Walls - A sample robot that demonstrates a wall-following movement pattern.
 * <p>
 * This robot navigates around the perimeter of the battlefield with the gun pointed inward.
 *
 * @author Mathew A. Nelson (original)
 * @author Flemming N. Larsen (contributor)
 */
public class Walls extends Robot {

	boolean peek; // Flag to determine if scanning is needed before turning
	double moveAmount; // The maximum distance to move along a wall

	/**
	 * run: Navigates the robot around the battlefield perimeter
	 */
	public void run() {
		// Set colors
		setBodyColor(Color.black);
		setGunColor(Color.black);
		setRadarColor(Color.orange);
		setBulletColor(Color.cyan);
		setScanColor(Color.cyan);

		// Set movement distance to the maximum battlefield dimension
		moveAmount = Math.max(getBattleFieldWidth(), getBattleFieldHeight());
		// Initialize scanning flag to false
		peek = false;

		// Align with a wall by turning to face it directly
		// Using modulo to calculate the angle needed to face a wall
		turnLeft(getHeading() % 90);
		ahead(moveAmount);
		// Enable scanning before turning
		peek = true;
		// Point gun toward the battlefield center
		turnGunRight(90);
		// Turn to move along the next wall
		turnRight(90);

		while (true) {
			// Enable scanning before moving along the wall
			peek = true;
			// Move along the current wall
			ahead(moveAmount);
			// Disable scanning during turning
			peek = false;
			// Turn 90 degrees to follow the next wall
			turnRight(90);
		}
	}

	/**
	 * onHitRobot: Handles collision with another robot by moving away
	 */
	public void onHitRobot(HitRobotEvent e) {
		// If the other robot is in front of us, move backward
		if (e.getBearing() > -90 && e.getBearing() < 90) {
			back(100);
		} // If the other robot is behind us, move forward
		else {
			ahead(100);
		}
	}

	/**
	 * onScannedRobot: Fires at the detected robot and performs additional scanning when needed
	 */
	public void onScannedRobot(ScannedRobotEvent e) {
		// Fire at the detected robot with power level 2
		fire(2);
		// Scan is automatically called during movement, but we need an additional scan
		// when the peek flag is true to detect robots before turning or moving toward a new wall
		if (peek) {
			scan();
		}
	}
}

File: plugins/dotnet/robocode.dotnet.samples/src/SampleCs/Walls.cs
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
using System;
using System.Drawing;
using Robocode;

namespace SampleCs
{
    /// <summary>
    ///   Walls - a sample robot by Mathew Nelson, and maintained by Flemming N. Larsen
    ///   <p />
    ///   Moves around the outer edge with the gun facing in.
    /// </summary>
    public class Walls : Robot
    {
        private bool peek; // Don't turn if there's a robot there
        private double moveAmount; // How much to move

        /// <summary>
        ///   run: Move around the walls
        /// </summary>
        public override void Run()
        {
            // Set colors
            BodyColor = (Color.Black);
            GunColor = (Color.Black);
            RadarColor = (Color.Orange);
            BulletColor = (Color.Cyan);
            ScanColor = (Color.Cyan);

            // Initialize moveAmount to the maximum possible for this battlefield.
            moveAmount = Math.Max(BattleFieldWidth, BattleFieldHeight);
            // Initialize peek to false
            peek = false;

            // turnLeft to face a wall.
            // getHeading() % 90 means the remainder of
            // getHeading() divided by 90.
            TurnLeft(Heading%90);
            Ahead(moveAmount);
            // Turn the gun to turn right 90 degrees.
            peek = true;
            TurnGunRight(90);
            TurnRight(90);

            while (true)
            {
                // Look before we turn when ahead() completes.
                peek = true;
                // Move up the wall
                Ahead(moveAmount);
                // Don't look now
                peek = false;
                // Turn to the next wall
                TurnRight(90);
            }
        }

        /// <summary>
        ///   onHitRobot:  Move away a bit.
        /// </summary>
        public override void OnHitRobot(HitRobotEvent e)
        {
            // If he's in front of us, set back up a bit.
            if (e.Bearing > -90 && e.Bearing < 90)
            {
                Back(100);
            } // else he's in back of us, so set ahead a bit.
            else
            {
                Ahead(100);
            }
        }

        /// <summary>
        ///   onScannedRobot:  Fire!
        /// </summary>
        public override void OnScannedRobot(ScannedRobotEvent e)
        {
            Fire(2);
            // Note that scan is called automatically when the robot is moving.
            // By calling it manually here, we make sure we generate another scan event if there's a robot on the next
            // wall, so that we do not start moving up it until it's gone.
            if (peek)
            {
                Scan();
            }
        }
    }
}